Update LocalCitationNetwork to newest version 1.27 for Cita

timwoelfle opened the pull request:
Dear Diego, Dominic and all Cita contributors

I hope you're well! I've recently stumbled upon the latest beta (1.0.0-beta.11) of Cita and have to say I'm impressed by the progress, congratulations to you and all collaborators! I think this release will have a lot of potential to gain traction.

I figured I'll use this opportunity to try to bring your included LocalCitationNetwork (LCN) fork up to date to future-proof it and to incorporate some new features. I created a new LCN fork (https://github.com/LocalCitationNetwork/LocalCitationNetwork.github.io/tree/cita) and merged your old fork into it (https://github.com/diegodlh/Local-Citation-Network/tree/cita). As I've changed the codebase quite a lot over the years, the merge-editor was useless. Instead I manually reviewed all of your commits from 2021 and integrated them in the correct places.

This is my best attempt of a direct merge of your fork and LCN's current version 1.27. I've included your original localization placeholders, even though terminology has slightly changed in LCN's new version (I've tried to follow current literature recommendations: https://doi.org/10.1136/bmj-2023-078384). Some new strings will need new placeholders. Maybe we can harmonize the terminology in the future. I also have some ideas for new features 🙂

Looking forward to hearing your thoughts! Best,

Tim

timwoelfle committed:
Removed Microsoft Academic option, which was discontinued in 2022

timwoelfle committed:
Fix live version (#3)

timwoelfle committed:
v0.98 with OpenAlex and Semantic Scholars support

+ Added support for OpenAlex (OA), the successor of Microsoft Academic, which brings back its high coverage, speed and abstract support. Also allows search by different id types (including PMID)
+ Open Alex supports highlighting retracted papers, e.g. when opening source doi:10.3390/en14227745 (flags can be missing though)
+ Added support for Semantic Scholar (S2), which is a bit slower than OA but allows easy implentation of "Outgoing suggestions". Also allows search by even more id types (including PMID, ARXIV)
* Deactivated OpenCitations (OC) option by default due to instability of its metadata API, it can still be activated through FAQ "Which APIs are supported?"
* Updated examples and FAQ to showcase new APIs
* Show API in tag labels
* Rotate citation network 90° clockwise when fullscreen toggled so widescreen space is used more efficiently
* Bumped Vue, Buefy and vis.js Network to newest versions
! Cannot highlight edges by clicking anymore
! Avoid duplicate counting of references for in-degree (e.g. https://api.crossref.org/works/10.7717/PEERJ.3544 has 5 references to DOI 10.1080/00031305.2016.1154108)
! Avoid showing Vue template before site is mounted with v-cloak

timwoelfle committed:
Small improvements

v0.99
* When individual API calls fail in OpenAlex / Semantic Scholar, it now returns all the remaining articles instead of failing completely
* Error messages are now messages (Buefy "toasts") instead of blocking alerts
* Always allow editing list of DOIs before importing
! Ordering of articles from a list of DOIs now works correctly even when these DOIs are lowercase or have trailing strings like "doi:" or "https://doi.org/" (e.g. coming from Zotero Cita)
! Always reload source in clickOpenReferences to avoid source not having citations property anymore and causing errors with Semantic Scholar

timwoelfle committed:
Fix padding of line-numbers in "Import list of DOIs"

timwoelfle committed:
Bugfixes

! Remove duplicate input articles not only for S2 but also other source APIs
! Fix bug in "Filter articles" function some articles are lacking a title

timwoelfle committed:
Update to 1.0

timwoelfle committed:
Fix example path

timwoelfle committed:
Updated to 1.1

timwoelfle committed:
Fixed URL for cache folder

timwoelfle committed:
Updated to 1.11
! Fixed data retrieval through OpenAlex (OA) (work.host_venue became deprecated).
! Fixed automatic 90° turning of Citation network when "fullscreen network" is toggled.
! Fixed experimental "Incoming suggestions" for OpenAlex (OA) (see v1.1)

timwoelfle committed:
Updated to 1.12
* Bumped Buefy and vis.js Network to newest versions.
! Fixed bug: Changing "Number shown" for "Incoming suggestions" and "Outgoing suggestions" didn't update networks anymore since 1.1.

timwoelfle committed:
v1.15
+ Retrieve "Incoming suggestions" also with OpenAlex (OA) now. Results in 1 extra API call per input article (disable by running "vm.getCitationsOA = false" in browser console).
+ Added "↻" / "↺" link to rotate the citation network by 90° and stopped automatic rotation when fullscreen is toggled as it did not always fit shape of network.
+ Added option to color-group author nodes in co-authorship network by either their first or last article. Show which article they're color-grouped by in author-node-tooltip.
* Network settings are now saved with each network (i.e. tab) separately and stored locally when "Autosave results locally" is on or a JSON of a network is downloaded (Citation network settings: "Node color", "Node size", "Show source article node", "↻" / "↺"; Co-authorship network settings: "Node color", "Show first names", "only authors with a minimum [2-10] input & suggested articles are shown"); consistent with "Number shown" for "Incoming suggestions" and "Outgoing suggestions" since v1.1.
* Adapted stabilization time for citation network according to its size: might take longer for larger networks but their layout ends up nicer now. Also added turning wheels to indicate the process of initializing networks.
* Stopped always initializing both citation and co-authorship networks at the same time, now only done the first time their tab is opened. Also stopped re-stabilizing after resizing.
* Reduced Crossref API load by leveraging select parameter (https://gitlab.com/crossref/issues/-/issues/511).
! Fixed overwriting of author ids through co-authorship network (this led to non-working author links to their OpenAlex / Semantic Scholar pages).

timwoelfle committed:
v1.16: Articles can also be filtered by author affiliation now

timwoelfle committed:
22.08.2023 v1.20
+ New API settings: The number of retrieved "Top references" (formerly "Incoming suggestions") and "Top citations" (formerly "Outgoing suggestions") can now be set: None / Top 10 (default) / Top 20 / Top 50 / All. The retrieval of all references / citations is fairly fast with OpenAlex (OA) & Semantic Scholar (S2) but slow with OpenCitations (OC) & Crossref (CR).
+ New Citation network settings: "Minimum degree references" and "Minimum degree citations" (default 2).
+ Data can now be downloaded in the RIS format in addtion to the CSV format
+ Tables can now be toggled to full screen. In this case the networks are deactived, increasing performance when large tables are shown.
+ Retrieve and save article type, publication date, volume, issue, firstPage, lastPage.
* Renaming: "Incoming suggestions" become "Top references" and "Outgoing suggestions" become "Top citations" for consistency with the academic literature and other tools (e.g. CitationChaser: https://estech.shinyapps.io/citationchaser/)
+ Link to Local citations and Local references in article table details (folding out).
+ Articles can also be filtered by article ID & DOI.
* "Top references" must not be referenced by at least 2 input articles anymore and "Top citations" must not be cited by at least 2 input articles anymore.
* Don't show edges between "Top references" / "Top citations" anymore so the number of incoming/outgoing edges (arrows) is equivalent to their in-degree and out-degree (also using Semantic Scholar's references/citations endpoints doesn't provide reference lists so the relations between them are unknown, only their relations to the input articles).
* Autosave now only saves up to 100 Top references & Top citations because of space constraints. Use "download full network: JSON" to save full network.
* Show loading-progress for "Top references" / "Top citations"
* Download JSON, CSV, RIS as blob, allowing much larger file sizes.
! Fix escaping of quotes in CSV exports.

timwoelfle committed:
v1.21
+ Tables are paginated now.
* Tables now show all retrieved references / citations (either "Top x" or "All"), the "number shown" setting was moved to the citation network settings as "Maximum references" and "Maximum citations".
* Co-authorship network now only shows authors of input articles.
* Removed comment in RIS export and always use type "JOUR" (journal article, compare https://en.wikipedia.org/wiki/RIS_(file_format)#Type_of_reference ) for better compatibility with EndNote.
* PMIDs don't need "pmid:" prefix anymore (supported by OpenAlex (OA) & Semantic Scholar (S2)).

timwoelfle committed:
v1.22
+ For Input Articles: Added "Filter References" and "Filter Citations" in article details.
+ For all articles: Added "Filter Input Articles cited by this" (corresponds to Out-Degree) and "Filter Input Articles citing this" (corresponds to In-Degree).
+ Added "Too Long Didn't Read" (TLDR) summaries for Semantic Scholar (S2).
* Use batch API for Semantic Scholar (S2), which allows much faster retrievals and much fewer API-overloads (error 429). (Unfortunately seems to have a bug, compare below "Known bugs")
* In case "All references" or "All citations" are retrieved (OA & S2 only): don't de-duplicate them anymore so that numbers in the respective tabs are correct, also when filtering them.
* Let users choose number of articles per page.
* Deactivated S2 reference contexts for now (simultaneous API call often led to API-overload (error 429)).
* Bumped Buefy, Vue and vis.js Network to newest versions. Use "render-on-mounted" for modals, introduced in new Buefy version, and removed previously necessary hack.
! Fixed "Show source article node" in Citation network settings.

Known bugs:
- Top Citations are wrong with Semantic Scholar (S2), they are pretty random citations and not the top ones. This seems to be a bug on their side (which now affects Local Citation Network because of using the batch API endpoint) which I have reported: https://github.com/allenai/s2-folks/issues/199

timwoelfle committed:
v1.23
+ Support Semantic Scholar API Keys (may occassionally speed up API, request one here: https://www.semanticscholar.org/product/api#api-key).
* Refactored code to make use of Object.freeze() for large arrays containing articles, increasing performance.
* Refactored code to calculate "referenced" and "citing" objects on the fly instead of caching them in the graph object. This reduces size of localStorage ("autosave results locally") and network json files ("download full network"). It leads to slightly more computation when changing the network-tab. Also made it more efficient by only looking at articles' references lists (instead of also citations lists) and thus reduced some redundancies.
! The above refactoring also fixed a bug that sometimes led to slightly wrong numbers of In-Degree and Out-Degree for References and Citations. This fix also applies retrospectively to saved networks after reloading.

timwoelfle committed:
v1.24
+ Now add placeholders in Input Articles table when id was set (e.g. in custom list of IDs or via bookmarklet) but not found via API or an error occurred.
+ Now truely retrieve All References / All Citations also with Semantic Scholar (S2) by implementing API paging, not just the first 1000 per paper.
* Optimized code to use recursive deepFreeze() instead of Object.freeze(), before actually adding article arrays/objects to Vue data tree, vastly increasing performance (especially noticable when All References & All Citations tables contain (tens of) thousands of rows).
* Refactored code to improve readability (splitting up the createNewNetwork function in createNewNetwork, retrievedInputArticles, retrievedIncomingSuggestions, retrievedOutgoingSuggestions).
* Added a few first end-to-end tests with cypress (more are needed for better maintainability).
* Updated examples to also include placeholders.

timwoelfle committed:
v1.25
* Remove Semantic Scholar (S2) responses without id (sometimes affected All References / All Citations).
! Fix formatting of '<' & '>' in titles.

timwoelfle committed:
Fixed filtering

timwoelfle committed:
06.10.2024 v1.26
+ Renamed "Input Articles" to "Seed Articles", "References" to "Cited", and "Citations" to "Citing" in accordance with "Guidance on terminology, application, and reporting of citation searching: the TARCiS statement" (Hirt 2024, https://doi.org/10.1136/bmj-2023-078384).
+ All Cited and All Citing are now de-duplicated against Seed Articles and All Citing also against All Cited by default (there is an option to turn this off).
+ Added CSV and RIS downloads for full network (i.e., all tabs).
+ RIS files now contain notes including numbers of Citations, References, Cited, Citing.
+ Saving vis.js network custom settings now using "download full network" and by turning Autosave on ("Save" in the top right). This means networks are more customizable now.
+ Some preparations for Co*Citation Network (compare https://doi.org/10.17605/OSF.IO/NPM2E)
- Removed Citation network option "Node size", now always default: Sum of In- & Out-Degree ("Rank").

timwoelfle committed:
v1.27
+ Export options modal for CSV / RIS download.
* De-duplication of All Cited and All Citing is off again by default but can be turned on.
! Fixed loading citation network settings "Max. number of Cited Articles" (maxCitedArticles) and "Max. number of Citing Articles" (maxCitingArticles) from old saves.

timwoelfle committed:
Merged https://github.com/diegodlh/Local-Citation-Network/tree/cita into current version of LCN (1.27) by manually resolving conflicts in index.html and index.js

timwoelfle committed:
Remove files not necessary fot Cita

timwoelfle committed:
Rename API from 'Cita' to 'Zotero Cita'

timwoelfle committed:
Remove unused tabs and hero-text

timwoelfle committed:
- Adds "download full network as JSON"
-- For sharing, can be reopened by anyone via "File" (top-right) on https://LocalCitationNetwork.github.io/
- Adds "export articles as CSV or RIS"
-- For consistency with online LCN and for downstream-use

timwoelfle committed:
- Do not de-duplicate citedArticlesIds against seedArticles, as LCN now shows all references in second-tab's table (called "All Cited" in web verison) and not just top 10
- Add datetime to the graph's tabLabel

timwoelfle committed:
- Show "Seed article: " in corresponding rows in Second Table (currently still called "Incoming Suggestions", name should be updated to "All Cited")
- Add column "#" showing "Number in custom list", corresponding to original order in Zotero (works through "numberInSourceReferences")
- "listOfDOIs" for Recreate network in web LCN now simply uses listOfKeys.map (instead of .reduce) to include gaps, which will keep numberInSourceReferences correct in web LCN
- Add customListOfReferences, which allows linkToShareAppendix to work properly also in web LCN.

timwoelfle committed:
v1.28
* Some adaptations for better integration with the awesome Zotero Cita plugin: https://github.com/diegodlh/zotero-cita.
* Slightly changed linkToShareAppendix() logic.
* Added downloadFile function.
! Fixed physics of Co-authorship network by reverting to default barnesHut configuration.
! Fix title of edges in Co-authorship network (only considers seed articles as of 1.21).
! Fixed some old terminology in comments.

timwoelfle committed:
Merge branch 'master' into cita

timwoelfle commented:
Today I've made some more changes, the most importing being the ability to download the network as a JSON file from within LCN. This allows users of Zotero Cita to easily share their networks with anyone, even non-Zotero users!

You just have to go to https://localcitationnetwork.github.io/ and open the JSON via "File" at the top right. I've attached an example file from myself. Hope you like this new feature, let me know what you think! Best,

Tim

[Zotero Cita 2025-02-02 17:04.json](https://github.com/user-attachments/files/18633351/Zotero.Cita.2025-02-02.17.04.json)

PS: I'm not experienced with GitHub pull requests, I hope I did everything correctly, otherwise please let me know! This new feature requires the saveFile function included in this pull request: https://github.com/diegodlh/zotero-cita/pull/313

timwoelfle committed:
Don't show Ref. column for  "All Cited" (="Incoming Suggestions")

Dominic-DallOsto commented:
Hey, sorry I missed this! I wasn't getting notifications from this repo.

This is amazing though, thanks a lot!

I'm actually not a maintainer of this repo so we'll have to wait for @diegodlh. But we can already point Cita to the cita branch [here](https://github.com/LocalCitationNetwork/LocalCitationNetwork.github.io/tree/cita)

Dominic-DallOsto commented:
If you have the time to quickly check out [this issue](https://github.com/diegodlh/zotero-cita/issues/262) too - that'd be amazing! Basically, can we make LCN within Cita more accessible via adaptive font sizes?

timwoelfle commented:
Hi Dominic, glad you like it! I'll try to check out the issue you've mentioned. It's too bad the Zotero-browser-window doesn't feature standard [Ctrl]+[+] / [Ctrl]+[-] shortcuts.

timwoelfle committed:
15.03.2025 v1.29
* Slightly increased standard font size from 9pt to 10pt for better readability.
* Made CSS layout more flexible by replacing most px measures in rem.
* Bumped vis.js Network to newest versions.
! Fixed bugmarklet, in rare cases decodeURIComponent(x.innerHTML) caused an "URIError: malformed URI sequence"

timwoelfle committed:
Merge LCN 1.29 into cita-branch

timwoelfle committed:
As Cita's LCN window doesn't support native browser-like zooming with [Ctrl]+[+] / [Ctrl]+[-] / [Ctrl]+[0], imitate it by adjusting document's fontSize (addresses https://github.com/diegodlh/zotero-cita/issues/262)

Dominic-DallOsto reviewed (commented):
Amazing - thanks a lot!

timwoelfle committed:
Don't show "Add network for references" buttons

timwoelfle committed:
- Show option to toggle de-duplication of Incoming Suggestions (=Cited Articels)
- Don't show Citing Articles in the "Export options" modal

timwoelfle committed:
Update internationalization (localization strings)

timwoelfle committed:
Clean index.html from parts unnecessary for Cita

timwoelfle committed:
Removed footer span

timwoelfle committed:
Fix Google Scholar links and minor cleanup (removal of code unsued in Cita)
